twisted/web/test/test_http2.py

```diff
@@ -1971,6 +1972,30 @@ def validate(streamID):
         return a._streamCleanupCallbacks[1].addCallback(validate)
 
 
+    def test_unnecessaryWindowUpdateForStream(self):
+        """
+        When a WindowUpdate frame is received for a stream but no data is
+        currently waiting, nothing exciting happens.
+        """
+        f = FrameFactory()
+        transport = StringTransport()
+        conn = H2Connection()
+        conn.requestFactory = DummyHTTPHandler
+
+        # Send the request.
+        frames = buildRequestFrames(
+            self.postRequestHeaders, self.postRequestData, f
+        )
+        frames.insert(1, f.buildWindowUpdateFrame(streamID=1, increment=5))
```

Comment:
So this high-level idea isn't bad, but it turns out to be a bit tricky in practice.

Firstly, at a low-level we already have those methods. hyperframe basically exposes those using the frame constructors.

The bigger problem is that there is connection state that needs to be managed. For example, HTTP/2 header compression is stateful, which means that we need somewhere to store a H2 header compressor that lasts for the duration of the test, and then pass it to all these helper methods (or, more likely, `serializeFrames`). The next problem is that both HEADERS and DATA frames also have stream IDs, and flags, all of which need to conditionally be populated depending on exactly how the stream state is being controlled.

We can certainly do this, but it doesn't vastly improve the state of affairs. Regardless, if we want to do this we should do it as part of a separate ticket, because all the tests in this file would want to be changed.